File: src/ts/magma/api/collect/head/FlatMapHead.ts
import { Head } from "../../../../magma/api/collect/head/Head";
import { Iter } from "../../../../magma/api/collect/Iter";
import { Option } from "../../../../magma/api/option/Option";
import { None } from "../../../../magma/api/option/None";
export class FlatMapHead<T, R> implements Head<R> {
	mapper: (arg0 : T) => Iter<R>;
	head: Head<T>;
	current: Iter<R>;
	constructor (head: Head<T>, initial: Iter<R>, mapper: (arg0 : T) => Iter<R>) {
		this.head/*unknown*/ = head/*Head<T>*/;
		this.current/*unknown*/ = initial/*Iter<R>*/;
		this.mapper/*unknown*/ = mapper/*(arg0 : T) => Iter<R>*/;
	}
	next(): Option<R> {
		while (true/*unknown*/){
			let next = this.current.next()/*unknown*/;
			if (next.isPresent()/*unknown*/){
				return next/*unknown*/;
			}
			let tuple = this.head.next().map(this.mapper).toTuple(this.current)/*unknown*/;
			if (tuple.left()/*unknown*/){
				this.current/*unknown*/ = tuple.right()/*unknown*/;
			}
			else {
				return new None<R>()/*unknown*/;
			}
		}
	}
}

File: src/magma/util/FlatMapHead.java
package magma.util;

import java.util.function.Function;

public class FlatMapHead<T, R> implements Head<R> {
    private final Head<T> head;
    private final Function<T, Iterator<R>> mapper;
    private Iterator<R> current;

    public FlatMapHead(Iterator<R> initial, Head<T> head, Function<T, Iterator<R>> mapper) {
        this.current = initial;
        this.head = head;
        this.mapper = mapper;
    }

    @Override
    public Option<R> next() {
        while (true) {
            final var maybeNext = current.next();
            if (maybeNext.isPresent()) {
                return maybeNext;
            }

            final var maybeNextIter = head.next().map(mapper);
            if (maybeNextIter.isPresent()) {
                current = maybeNextIter.get();
            }
            else {
                return new None<>();
            }
        }
    }
}
